### PR TITLE
Change key file config

### PR DESCRIPTION
**Pull request by CodingParsley**
What
----
Add configuration for passing BigQuery credential key as json string instead of file path since some cloud infrastructure does not support storing external credential files.

Solution: 
Add keyfile.isJsonString boolean configuration. When the field is configured and is true, the connector will interpret the field value as a json string.

Test&Review
------------
credential interpreted correctly, does not break BigQuery connector

This PR also has changes in PR #3 

What
----
1. Filter messages
Add configuration for message filtering to the BigQuery connector. When the fields "includeCondition" and "excludeCondition" are configured, the connector will decide if it should include or drop out the message based on the conditions provided.

Solution: 
Use JsonPath format to specify conditions in the configuration, and use the api to filter messages based on values of fields with specified jsonpath.

2. Sanitize field names
BigQuery dataset only allows underscore as special character for field names, and field names have to start with a letter or underscore.

3. Consume raw json in addition to avro
kafka connect json converter maps json to connector as sinkrecords. With modification of sinkrecord converter, the connector is able to consume json data without schema.

References
----------

https://github.com/json-path/JsonPath
https://cloud.google.com/bigquery/docs/schemas#column_names

Test&Review
------------
Wrote Junit tests.

Sample configurations:
https://confluentinc.atlassian.net/wiki/spaces/OBSERVE/pages/920783106/Sample+Configurations+for+record+filter+and+field+sanitizer

@xvrl @jcrobak @avocader

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -195,20 +202,37 @@ project(':kcbq-connector') {
...
+
+                "ch.qos.logback:logback-classic:$logbackVersion",
+                "org.junit.jupiter:junit-jupiter-api:$junitjupiterVersion-M2",
+                "org.junit.platform:junit-platform-runner:$junitPlatformVersion-M4",
                 "org.mockito:mockito-core:$mockitoVersion"
         )
+
+        testRuntime (
+                "org.junit.vintage:junit-vintage-engine:$junitVintageVersion-M4",
+                "org.junit.jupiter:junit-jupiter-engine:$junitjupiterVersion-M4",
```
Comment (jcrobak) [inline build.gradle]: it doesn't look like we need to add these deps, from what I see...

File: build.gradle [hunk 2]
```diff
@@ -277,7 +302,7 @@ project('kcbq-confluent') {
     configurations.all {
         resolutionStrategy {
             // depending on debezium results in us pulling in the most recent kafka version, but we want to use 1.1.1.
```
Comment (jcrobak) [inline build.gradle]: need to update this comment / or figure out if we can stick with 1.1.1.

File: build.gradle [hunk 3]
```diff
@@ -293,15 +318,21 @@ project('kcbq-confluent') {
...
+        mavenLocal()
         mavenCentral()
         maven {
             url 'http://packages.confluent.io/maven'
         }
+        maven {
+            url 'http://github.com/FasterXML/jackson'
+        }
+        maven { url 'http://kafka.apache.org' }
+        maven { url "http://repo.typesafe.com/typesafe/releases/" }
```
Comment (jcrobak) [inline build.gradle]: Are these extra maven repos necessary?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQueryHelper.java [hunk 4]
```diff
@@ -67,6 +68,44 @@ public BigQuery connect(String projectName, String keyFilename) {
...
+   * from specific datasets.
+   *
+   * @param projectName The name of the BigQuery project to work with
+   * @param keyFile The name of a file containing a JSON key that can be used to provide
+   *                    credentials to BigQuery, or null if no authentication should be performed.
+   * @param isJsonString If isJsonString is set to true, the method will interpret the keyFile as
+   *                     JsonString instead of file path
+   * @return The resulting BigQuery object.
+   */
+  public BigQuery connect(String projectName, String keyFile, Boolean isJsonString) {
```
Comment (jcrobak) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQueryHelper.java]: probably should use `boolean` so that the value can be `true` or `false` but not `null`.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java [hunk 5]
```diff
@@ -44,24 +44,18 @@
...
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.time.Instant;
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.Map;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
```
Comment (jcrobak) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java]: it doesn't look like this project uses wildcard imports, so you should probably undo this change.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java [hunk 6]
```diff
@@ -83,6 +77,7 @@
   private boolean useMessageTimeDatePartitioning;
 
   private TopicPartitionManager topicPartitionManager;
+  private SinkRecordFilter sinkRecordFilter;
```
Comment (jcrobak) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java]: did you mean to put the filter logic in this PR?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/SimpleBigQueryWriter.java [hunk 7]
```diff
@@ -62,8 +62,13 @@ public SimpleBigQueryWriter(BigQuery bigQuery, int retry, long retryWait) {
                                                             List<InsertAllRequest.RowToInsert> rows,
                                                             String topic) {
     InsertAllRequest request = createInsertAllRequest(tableId, rows);
+    logger.info("rows to insert: {}", rows);
     InsertAllResponse writeResponse = bigQuery.insertAll(request);
     if (writeResponse.hasErrors()) {
+      for (Map.Entry<Long, List<BigQueryError>> entry : writeResponse.getInsertErrors().entrySet()) {
+        logger.info("{}", entry.getValue());
+      // inspect row error
+      }
```
Comment (jcrobak) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/SimpleBigQueryWriter.java]: these look like leftover debug statements that should be removed.

File: build.gradle (hunk 1 above)
Comment (CodingParsley) [reply build.gradle]: Sorry it's confusing... This pr is branched off based on the filter/sanitize pr and include the changes in that pr...

File: build.gradle [hunk 8]
```diff
@@ -195,20 +202,37 @@ project(':kcbq-connector') {
...
+                "io.confluent:kafka-connect-avro-converter:$ioConfluentVersion",
+                "io.confluent:kafka-schema-registry-client:$ioConfluentVersion",
+                "org.apache.avro:avro:$avroVersion",
                 "org.apache.kafka:connect-api:$kafkaVersion",
                 "org.apache.kafka:kafka-clients:$kafkaVersion",
                 "org.slf4j:slf4j-api:$slf4jVersion",
                 "org.slf4j:slf4j-simple:$slf4jVersion",
+                "com.jayway.jsonpath:json-path:$jsonPathVersion",
+                "ch.qos.logback:logback-core:$logbackVersion",
+                "log4j:log4j:$log4jVersion"
```
[the remark this answers is not in the document]
Comment (CodingParsley) [reply build.gradle]: Sorry about the confusion...This pr is based off the branch filter&sanitize so the changes in the other pr are also included here...

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQueryHelper.java (hunk 4 above)
Comment (CodingParsley) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQueryHelper.java]: The BigQuerySinkConfig class will transform boolean properties to Boolean

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java (hunk 5 above)
Comment (CodingParsley) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java]: Seems the change is from mainstream and is used in the file

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java (hunk 6 above)
Comment (CodingParsley) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java]: Sorry about the confusion... This PR is branched off the filter&sanitize branch and has changes in the other PR.

File: build.gradle (hunk 2 above)
Comment (CodingParsley) [reply build.gradle]: @avocader Can I update the comment?

File: build.gradle [hunk 9]
```diff
@@ -195,20 +202,37 @@ project(':kcbq-connector') {
...
+                "com.google.cloud:google-cloud-bigquery:$googleCloudVersion",
+                "com.google.cloud:google-cloud-storage:$googleCloudVersion",
                 "com.google.auth:google-auth-library-oauth2-http:$googleAuthVersion",
-                "com.google.code.gson:gson:$googleCloudVersion",
+                "com.google.code.gson:gson:$googleCloudGsonVersion",
                 "io.debezium:debezium-core:$debeziumVersion",
+
+                "io.confluent:kafka-connect-avro-converter:$ioConfluentVersion",
+                "io.confluent:kafka-schema-registry-client:$ioConfluentVersion",
+                "org.apache.avro:avro:$avroVersion",
```
[the remark this answers is not in the document]
Comment (jcrobak) [reply build.gradle]: let's look at this again.

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/preprocess/SinkRecordFilterTest.java [hunk 10]
```diff
@@ -0,0 +1,198 @@
+package com.wepay.kafka.connect.bigquery.preprocess;
+
+import com.jayway.jsonpath.InvalidPathException;
+import com.wepay.kafka.connect.bigquery.SinkConnectorPropertiesFactory;
+import com.wepay.kafka.connect.bigquery.config.BigQuerySinkConfig;
+import org.apache.log4j.AppenderSkeleton;
+import org.apache.log4j.Logger;
```
Comment (jcrobak) [inline kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/preprocess/SinkRecordFilterTest.java]: Should use `org.slf4j.Logger` and `org.slf4j.LoggerFactory`

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/preprocess/SinkRecordFilter.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
+package com.wepay.kafka.connect.bigquery.preprocess;
+
+import com.jayway.jsonpath.*;
+import com.wepay.kafka.connect.bigquery.config.BigQuerySinkConfig;
+import com.wepay.kafka.connect.bigquery.convert.RecordConverter;
+import org.apache.kafka.common.config.ConfigException;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.apache.log4j.Logger;
```
Comment (jcrobak) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/preprocess/SinkRecordFilter.java]: Should use `org.slf4j.LoggerFactory` instead and remove the dependency from build.gradle.

File: build.gradle [hunk 12]
```diff
@@ -9,14 +9,20 @@ project.ext {
     apacheHttpClientVersion = '4.5.6'
     avroVersion = '1.8.1'
     debeziumVersion = '0.6.1'
-    googleCloudVersion = '0.47.0-alpha'
+    googleCloudVersion = '1.79.0'
     googleAuthVersion = '0.9.0'
     googleCloudGsonVersion = '2.8.5'
+    logbackVersion = '1.2.3'
```
Comment (jcrobak) [inline build.gradle]: definitely don't need both logback and log4j. You can only have one at a time.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/preprocess/SinkRecordFilter.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+            try {
+                // return a list map nodes that satify the specified condition
+                filterList = this.jsonPathCondition.read(
+                        convertedRecord,
+                        Configuration.defaultConfiguration().addOptions(
+                                Option.ALWAYS_RETURN_LIST
+                        )
+                );
+            } catch (PathNotFoundException e) {
+                logger.info("Json Path " + this.configString + " specified in " + this.configType +
```
Comment (jcrobak) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/preprocess/SinkRecordFilter.java]: maybe this should be debug logging

**Comment by xvrl** [comment]
looks like I've been reviewing #3 instead of this PR, which one is the right one :) ?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/preprocess/SinkRecordFilter.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+                        configType + "config is incorrectly formatted. " +
+                        "Please refer to com.jayway.jsonpath java doc for correct use of jsonpath.");
+            }
+        }
+
+    }
+
+
+    public List getFilterFields(Map<String, Object> convertedRecord) {
+        List filterList = new ArrayList();
```
Comment (jcrobak) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/preprocess/SinkRecordFilter.java]: add in template/type info.

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/preprocess/SinkRecordFilterTest.java (hunk 10 above)
Comment (CodingParsley) [reply kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/preprocess/SinkRecordFilterTest.java]: Done

File: build.gradle [hunk 15]
```diff
@@ -316,13 +335,16 @@ project('kcbq-confluent') {
         compile (
                 project(':kcbq-api'),
 
-                "com.google.cloud:google-cloud:$googleCloudVersion",
+                "com.google.cloud:google-cloud-bigquery:$googleCloudVersion",
```
Comment (CodingParsley) [inline build.gradle]: These are upstream changes

File: gradle/wrapper/gradle-wrapper.properties [hunk 16]
```diff
@@ -1,6 +1,6 @@
-#Tue Jul 05 08:29:00 PDT 2016
+#Fri Jul 12 08:28:50 PDT 2019
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-2.13-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-2.13-all.zip
```
Comment (CodingParsley) [inline gradle/wrapper/gradle-wrapper.properties]: upstream change

File: gradle/wrapper/gradle-wrapper.properties [hunk 17]
```diff
@@ -1,6 +1,5 @@
-#Tue Jul 05 08:29:00 PDT 2016
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.8.1-bin.zip
```
Comment (CodingParsley) [inline gradle/wrapper/gradle-wrapper.properties]: update gradle version

File: gradlew [hunk 18]
```diff
@@ -1,4 +1,4 @@
-#! /usr/bin/env bash
```
Comment (CodingParsley) [inline gradlew]: update gradlew version

File: build.gradle (hunk 9 above)
[the remark this answers is not in the document]
Comment (CodingParsley) [reply build.gradle]: Done

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/preprocess/SinkRecordFilter.java (hunk 14 above)
Comment (CodingParsley) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/preprocess/SinkRecordFilter.java]: like in logs?

File: build.gradle [hunk 19]
```diff
@@ -233,7 +244,7 @@ project('kcbq-api') {
     jar {
         manifest {
             attributes  'Implementation-Title': 'Kafka Connect BigQuery API',
-                        'Implementation-Version': version
```
Comment (CodingParsley) [inline build.gradle]: Wrong indentation
